### PR TITLE
tpm2_listpcrs: use TPM2_GetCapability to determine PCRs to read

### PR DESCRIPTION
**Pull request by snits**
Allow TPM to tell us PCRs that are available to be read in a bank
instead of assuming there will be 24. This resolves an issue on
a system where in the bios you can choose between hash functions
(sha1 and sha256 in this case). Both will show up as supported,
but when it tries reading the bank that isn't selected in the bios
it makes no progress, and then fails saying that there are too many
PCRs to read.

This consolidates the pcr_selections init code into one function.

Also a minor change to show_pcr_values to use sizeofSelect to
determine loop iteration instead of hardcoding 24.

Signed-off-by: Jerry Snitselaar <jsnitsel@redhat.com>

### CODE REVIEW

**Comment by snits** [comment]
Hi William,
I'm still trying to decide if this is the best way to handle this. I'm trying to figure out a decent way to push this down into get_banks, since TPM2_GetCapability is already being called there. As mentioned in the commit message, we've seen this problem where on certain systems you can select a hash function in the bios for the tpm. Both hash functions show up as supported with tpm2_listpcrs -s, but when it tries to read the function that isn't selected in the bios the pcr read calls succeed, but there is nothing there so the bitfields are getting cleared out. Eventually it iterates through 24 times, jumps out of the loop, sees that it still has pcrs to read, and complains about it.

That CI run failed because it couldn't find realpath. Not sure what happened there.

File: tools/tpm2_listpcrs.c [hunk 1]
```diff
@@ -173,38 +173,36 @@ static bool read_pcr_values(listpcr_context *context) {
...
-    pcr_selections->count = 1;
-    pcr_selections->pcrSelections[0].hash = alg_id;
-    set_pcr_select_size(&pcr_selections->pcrSelections[0], 3);
-    clear_pcr_select_bits(&pcr_selections->pcrSelections[0]);
-
-    UINT32 pcr_id;
-    for (pcr_id = 0; pcr_id < 24; pcr_id++) {
-        set_pcr_select_bit(&pcr_selections->pcrSelections[0], pcr_id);
+static bool init_pcr_selection(TPMI_ALG_HASH alg_id, listpcr_context *context)
+{
```
Comment (williamcroberts) [inline tools/tpm2_listpcrs.c]: Currently this code base is a mixed mess of style, but adhere to the coding style under: tpm2.0-tools/misc/coding_standard_c.txt.

Functions have opening brace on the same line, variables use underscores, no camel case.

**Comment by williamcroberts** [comment]
Thank you for this, I think this takes care of: https://github.com/01org/tpm2.0-tools/issues/119 and I miss diagnosed how to handle that issue.

**Comment by snits** [comment]
Still failing to find realpath.

Sorry about the style. Hopefully it is good now. I'm dealing with 1.1.0 and this at the same time so get confused at times.

**Comment by williamcroberts** [comment]
Dont be sorry, we like to see changes, ill look into the Travis issue.

**Comment by williamcroberts** [comment]
Hopefully PR #285 fixes it.

**Comment by williamcroberts** [comment]
That's actually not the problem. We just set PATH in the .travis.yml file. The problem is, that the resourcemanager in TSS has been removed, so were not actually starting the resourcemanager. So we need to actually grab the 1.0 branch, or better yet swicth to the new resource manager.

However, this highlights some issues in the test script, we have a test passing!

**Comment by williamcroberts** [comment]
Sorry for the delay, can you re-push. I fixed the travis build.

File: tools/tpm2_listpcrs.c [hunk 2]
```diff
@@ -271,14 +268,16 @@ static bool show_selected_pcr_values(listpcr_context *context) {
 
 static bool show_all_pcr_values(listpcr_context *context) {
 
-    init_pcr_selection_all(&context->algs, &context->pcr_selections);
+    if (!init_pcr_selection(0, context))
```
Comment (williamcroberts) [inline tools/tpm2_listpcrs.c]: normally we don't like to bury routines in conditionals, but ill let this slide.

**Comment by williamcroberts** [comment]
Nevermind on re-pushing, I found a restart build button on Travis...

**Comment by williamcroberts** [comment]
Moved to: https://github.com/01org/tpm2.0-tools/pull/286

**Pull request closed without merging**